Mathlib/Algebra/Homology/DerivedCategory/Ext.lean

```diff
@@ -0,0 +1,145 @@
+    HasExt.{w} C ↔ ∀ (X Y : C) (n : ℤ), Small.{w}
+      ((DerivedCategory.singleFunctor C 0).obj X ⟶
+        (((DerivedCategory.singleFunctor C 0).obj Y)⟦n⟧)) := by
+  dsimp [HasExt]
+  simp only [hasSmallLocalizedShiftedHom_iff _ _ DerivedCategory.Q]
+  constructor
+  · intro h X Y n
+    exact (small_congr ((shiftFunctorZero _ ℤ).app
+      ((DerivedCategory.singleFunctor C 0).obj X)).homFromEquiv).1 (h X Y 0 n)
+  · intro h X Y a b
+    refine (small_congr ?_).1 (h X Y (b - a))
+    exact (Functor.FullyFaithful.ofFullyFaithful
+      (shiftFunctor _ a)).homEquiv.trans
+      ((shiftFunctorAdd' _ _ _ _ (Int.sub_add_cancel b a)).symm.app _).homToEquiv
+
+lemma hasExt_of_hasDerivedCategory [HasDerivedCategory.{w} C] : HasExt.{w} C := by
+  rw [hasExt_iff.{w}]
+  infer_instance
+
+variable {C}
+
+variable [HasExt.{w} C]
+
+namespace Abelian
+
+/-- A Ext-group in an abelian category `C`, defined as a `Type w` when `[HasExt.{w} C]`. -/
+def Ext (X Y : C) (n : ℕ) : Type w :=
+  SmallShiftedHom.{w} (HomologicalComplex.quasiIso C (ComplexShape.up ℤ))
+    ((CochainComplex.singleFunctor C 0).obj X)
+    ((CochainComplex.singleFunctor C 0).obj Y) (n : ℤ)
+
+namespace Ext
+
+variable {X Y Z T : C}
+
+/-- The composition of `Ext`. -/
+noncomputable def comp {a b : ℕ} (α : Ext X Y a) (β : Ext Y Z b) {c : ℕ} (h : a + b = c) :
+    Ext X Z c :=
+  SmallShiftedHom.comp α β (by omega)
+
+lemma comp_assoc {a₁ a₂ a₃ a₁₂ a₂₃ a : ℕ} (α : Ext X Y a₁) (β : Ext Y Z a₂) (γ : Ext Z T a₃)
+    (h₁₂ : a₁ + a₂ = a₁₂) (h₂₃ : a₂ + a₃ = a₂₃) (h : a₁ + a₂ + a₃ = a) :
+    (α.comp β h₁₂).comp γ (show a₁₂ + a₃ = a by omega) =
+      α.comp (β.comp γ h₂₃) (by omega) :=
+  SmallShiftedHom.comp_assoc _ _ _ _ _ _ (by omega)
+
+section
+
+variable [HasDerivedCategory.{w'} C]
+
+/-- When an instance of `[HasDerivedCategory.{w'} C]` is available, this is the bijection
+between `Ext.{w} X Y n` and a type of morphisms in the derived category. -/
+noncomputable def homEquiv {n : ℕ} :
+    Ext.{w} X Y n ≃ ShiftedHom ((DerivedCategory.singleFunctor C 0).obj X)
+      ((DerivedCategory.singleFunctor C 0).obj Y) (n : ℤ) :=
+  SmallShiftedHom.equiv (HomologicalComplex.quasiIso C (ComplexShape.up ℤ)) DerivedCategory.Q
+
+/-- The morphism in the derived category which corresponds to an element in `Ext X Y a`. -/
+noncomputable abbrev hom {a : ℕ} (α : Ext X Y a) :
+    ShiftedHom ((DerivedCategory.singleFunctor C 0).obj X)
+      ((DerivedCategory.singleFunctor C 0).obj Y) (a : ℤ) :=
+  homEquiv α
+
+@[simp]
+lemma comp_hom {a b : ℕ} (α : Ext X Y a) (β : Ext Y Z b) {c : ℕ} (h : a + b = c) :
+    (α.comp β h).hom = α.hom.comp β.hom (by omega) := by
+  apply SmallShiftedHom.equiv_comp
+
+@[ext]
+lemma ext {n : ℕ} {α β : Ext X Y n} (h : α.hom = β.hom) : α = β :=
+  homEquiv.injective h
+
+end
+
+-- TODO: promote `homEquiv` above as `homAddEquiv` with the same
+-- degree of generality, i.e. for any `[HasDerivedCategory.{w'} C]`
+noncomputable instance {n : ℕ} : AddCommGroup (Ext X Y n) :=
```

Comment:
I'm a bit confused. How can Lean find the right `w'` to get a `HasDerivedCategory.{w'} C` instance?
Or does Lean ignore the universe parameter and use the first `HasDerivedCategory.{*} C`? Do we need compatibility lemmas for multiple `HasDerivedCategory.{*} C` instances?

Reply:
The idea is that the definition of `Ext` groups does not depend on any `HasDerivedCategory` variable. For the definition of the abelian group structure, the best solution I have found is to define it using the constructed derived category. Then, using #14153 and some more (painful) work (wip PR #13989), I shall show that the abelian group structure is compatible with any `HasDerivedCategory` variable. At the same time, we shall also see that an exact functor induces maps on `Ext`-groups that are compatible with `Functor.mapDerivedCategory`. (By design, I believe we should not define any global instance of `HasDerivedCategory`. It is like with `Category` for which the universe is determined by the first (and hopefully unique) variable found.)

Reply:
By "constructed derived category", do you mean the "formally invert all arrows" large universe construction?
I think it would be nicer to specify which `HasDerivedCategory` is used explicitly here rather than relying on the belief that "we should not define any global instance of `HasDerivedCategory`" in order to avoid nasty diamonds, but I IIRC we don't have that construction in mathlib yet right?

Reply:
Yes, the "constructed derived category" is the large universe construction (this is the target category of `MorphismProperty.Q`), which is already used *internally* in order to define `Ext`. This is the same localized category that is provided by `HasDerivedCategory.standard`.

Providing global instances of `HasDerivedCategory` for general categories is a very bad idea, because it would mean that we would basically have to specify the universes everywhere when we use the `DerivedCategory` API...

The only other reasonable approach I see would be to define the `Add` instance on the `Ext` types using suitable binary biproducts and the basic functoriality of `Ext`. I have just removed the `AddCommGroup` instance, and we see the details in future PRs.

Reply:
Ah sorry I think I misread the original approach (I thought it took a `[HasDerivedCategory.{w} C]` instance variable) and that approach is perfectly fine IMO.

Reply:
The other approach I am suggesting just above might be slightly more convenient, I will experiment a little bit with it. The better is to keep the `AddCommGroup` instance out of this PR.